src/middleware/middleware-login.ts

```diff
@@ -0,0 +1,34 @@
+import { Request, Response, NextFunction } from 'express';
+import HttpStateCodes from '../utils/http-state-codes';
+
+export const LoginMiddleware = (req: Request, res: Response, next: NextFunction) => {
+  const { email, password } = req.body
+  
+  const expectedFields = ['email','password'];
+  const validKeys = Object.keys(req.body);
+
+  const missingFields = expectedFields.filter((field) => !validKeys.includes(field));
+
+  const extraFields = validKeys.filter((key) => !expectedFields.includes(key));
+
+  if (missingFields.length > 0) {
+    return res.json({
+      message:`Faltan los siguientes campos en la solicitud: ${missingFields.join(', ')}`, 
+      status:HttpStateCodes.BAD_REQUEST});
+  }
+
+  if (extraFields.length > 0) {
+    return res.json({message:`Campos no válidos en la solicitud: ${extraFields.join(', ')}`, status:HttpStateCodes.BAD_REQUEST});
+ }
+
+if(!(/^[^\s@]+@[^\s@]+\.[^\s@]{2,20}$/).test(email)){
```

Comment:
Ok this is fine but not the correct way to do validation. I haven't explained this part in classes so  really sorry.
Ok I will explain here the correct way to do a perfect validation.
There are several libraries and approaches available for this purpose, each with its own strengths. Joi is indeed one of the popular choices due to its powerful schema description language and data validation capabilities. However, the "best" way can depend on your specific needs and preferences. Here are a few options you might consider:
**Joi** 
[https://www.npmjs.com/package/joi](https://www.npmjs.com/package/joi)

Joi provides a rich set of functionalities for validating JavaScript objects. It's especially powerful for complex validations and has good support for custom error messages.
you can improve this code using Joi and do a perfect separation between logic and schemas.
In a different file you can create the schemas, please create the schema in Joi and in TypeScript 
```
  const loginSchema = Joi.object({
    email: Joi.string().email().required().messages({
      'string.email': loginMessage.invalidEmail,
      'any.required': loginMessage.noEmail
    }),
    password: Joi.string().pattern(new RegExp('^(?=.*[A-Za-z])(?=.*\\d)[A-Za-z\\d]{8,20}$')).required().messages({
      'string.pattern.base': 'some messages from the loginMessages Object',
      'any.required': 'some messages from the loginMessages Object'
    })
  });
  exports authorSchema;
```
After that, you need to import this schema in your validation logic file.
```
export const ValidateFields = (req: Request, res: Response, next: NextFunction) => {
  // Define the schema

  // Validate the request body against the schema
  const { error } = schema.validate(req.body, { abortEarly: false });

  // Check for validation errors
  if (error) {
    // Format error messages
    const messages = error.details.map(detail => detail.message).join(", ");

    return res.status(HttpStateCodes.BAD_REQUEST).json({
      message
    });
  }

  next();
};
```